src/main/scala/org/neo4j/spark/reader/Neo4jDataSourceReader.scala

```diff
@@ -19,20 +21,43 @@ class Neo4jDataSourceReader(private val options: DataSourceOptions, private val
   private val neo4jOptions: Neo4jOptions = new Neo4jOptions(options.asMap())
     .validate(options => Validations.read(options, jobId))
 
-  override def readSchema(): StructType = {
-    val schemaService = new SchemaService(neo4jOptions, jobId)
+
+  override def readSchema(): StructType = callSchemaService { schemaService => schemaService
+    .struct() }
+
+  private def callSchemaService[T](lambda: SchemaService => T): T = {
+    val driverCache = new DriverCache(neo4jOptions.connection, jobId)
+    val schemaService = new SchemaService(neo4jOptions, driverCache)
+    var hasError = false
     try {
-      schemaService.struct()
+      lambda(schemaService)
+    } catch {
+      case e: Throwable => {
+        hasError = true
+        throw e
+      }
     } finally {
       schemaService.close()
+      if (hasError) {
+        driverCache.close()
+      }
     }
   }
 
+  private def getPartitions(): Seq[(Long, Long)] = callSchemaService { schemaService => schemaService
+    .skipLimitFromPartition() }
+
   override def planInputPartitions: util.ArrayList[InputPartition[InternalRow]] = {
     val schema = readSchema()
-    val factoryList = new java.util.ArrayList[InputPartition[InternalRow]]
-    factoryList.add(new Neo4jInputPartitionReader(neo4jOptions, filters, schema, jobId))
-    factoryList
+    val partitionSkipLimit = getPartitions()
+    val neo4jPartitions = if (partitionSkipLimit.isEmpty) {
+      Seq(new Neo4jInputPartitionReader(neo4jOptions, filters, schema, jobId))
+    } else {
+      partitionSkipLimit.zipWithIndex
```

Comment:
this code is not very readable, can we use a caseclass instead of the triple?

Reply:
Done